Sources/NIOCore/NIOLoopBound.swift

```diff
@@ -52,6 +60,23 @@ public struct NIOLoopBound<Value>: @unchecked Sendable {
             self._value = newValue
         }
     }
+
+    /// Access the `value` with the assertion that the code is running on `eventLoop`.
+    ///
+    /// Unlike ``NIOLoopBound/value``, this performs the assertion in debug configuration only, so it's
+    /// cheaper, and still performs the precondition check in debug mode.
+    /// - note: ``NIOLoopBound`` itself is value-typed, so any writes will only affect the current value.
+    @inlinable
+    public var uncheckedValue: Value {
+        get {
+            self._eventLoop.assertInEventLoop()
+            return self._value
+        }
+        set {
+            self._eventLoop.assertInEventLoop()
+            self._value = newValue
+        }
+    }
```

Comment:
What if we just provide a second type that does asserts or preconditions. Alternatively what about just making the whole precondition/assert behaviour configurable on init?

Reply:
I think this pattern is better: I don't see a good reason to make this a type or a runtime distinction. It's good to be able to see this in the code, as it's fundamentally a safety-breaking optimization.

Reply:
TBH I'm also thinking a type would be better. It's how the rest of Swifts unsafe/unchecked APIs usually work e.g. `UncheckedContinuation`, `Unsafe*Pointer`, `@unchecked Sendable`.

It may also enable us to drop the stored `EventLoop` property in release modes, once `DEBUG` is properly propagated in SwiftPM (haven't checked recently), making this actually a zero cost abstraction.

Reply:
One type is better here: I often want mixed use: First time in a function: checked, follow-ups: unchecked. Different type would be pain and would likely push me to use unchecked everywhere.

Reply:
> `First time in a function: checked, follow-ups: unchecked`

It would be better to get the value checked out of the loop bound and just store it in a local `let` for future accesses.

Reply:
> It's how the rest of Swifts unsafe/unchecked APIs usually work

This is not true: the number of _safe_ types with _unsafe_ functions vastly dwarfs the number of unsafe types.

Reply:
> the number of safe types with unsafe functions vastly dwarfs the number of unsafe types

All the `withUnsafe*` functions I can think off vend you types that have `Unsafe` or `Unchecked` in the name. About which functions do you think?
I'm only aware of `[Closed]Range(uncheckedBounds:)` which result in a type that doesn't have `Unsafe` in the name.

Reply:
Concretely, I'm proposing this API:
```swift
public struct NIOUncheckedLoopBound<Value> {
    public init(_ value: Value, eventLoop: EventLoop) {}
    public init(_ value: NIOLoopBound<Value>) {}
    public init(_ value: NIOLoopBoundBox<Value>) {}
    
    public var value: Value {
        get {}
    }
}

public final class NIOUncheckedLoopBoundBox<Value> {
    public init(_ value: Value, eventLoop: EventLoop) {}
    public init(_ value: NIOLoopBound<Value>) {}
    public init(_ value: NIOLoopBoundBox<Value>) {}
    
    public var value: Value {
        get {}
        set {}
    }
}
```

If `NIOLoopBound.value` is to expensive you can still wrap it in `NIOUncheckedLoopBound(myLoopBoundValue).value`. Its more verbose but that's a good thing.

Reply:
Whoa, thank you, @dnadoba, @Lukasa, and @weissi for the discussion! I'm learning quite a lot from you here.

I'm not attached to the current implementation — as I learn more and if we decide to change it, I'm happy to rewrite it.

> You can't say in isolation if loopBound.uncheckedValue is safe or not. You need to go and look at the creation of the NIOLoopBound and trace back on which EventLoop it was created. 

I agree, and I agree it's different from the `Range` example.

> We are proposing a set of duplicated APIs which make all operations on NIOLoopBound unsafe/unchecked.
> This is more in line with the UnsafePointer APIs where almost all operation are unsafe.

> This design also gives the wrong impression that NIOLoopBound.value is always safe and only uncheckedValue is unsafe. With this PR value becomes unsafe as well as you might have initialised it with NIOLoopBound(_:uncheckedEventLoop).

If we choose to stick with the current implementation, I think I should at the very least comment and document on that behavior.

> Such a type is borderline useless, because to support people using it you have to litter it all over your codebase.

@Lukasa, with the `Range` example, I'd fully agree. For `NIOLoopBound`, though — we're not passing `LoopBound` around in the NIO codebase right now, right? So in theory, sure, we could do two separate structures. The weight of supporting two different structures would then be on the user. Perhaps we could provide a `NIOLoopBound` as a protocol, and make a safe and unchecked implementations of that protocol, so that users could accept `some NIOLoopBound` and deal with them that way?


Again, I feel that I'm out of my depth in the API design discussion, but I'm delighted to be part of it and listen in. Feel free to "well actually" me on any points I'm bringing up.